From a3e5aa9f264338b7043dedd7e4a049be3e137953 Mon Sep 17 00:00:00 2001
From: Jan Janssen <jan-janssen@users.noreply.github.com>
Date: Thu, 27 Feb 2025 20:42:41 +0100
Subject: [PATCH] ElectronicStructureClass (#29)

* ElectronicStructureClass

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
---
 pyiron_vasp/vasp/output.py  | 13 ++++++++++---
 pyiron_vasp/vasp/vasprun.py |  4 ++--
 2 files changed, 12 insertions(+), 5 deletions(-)

diff --git a/pyiron_vasp/vasp/output.py b/pyiron_vasp/vasp/output.py
index e778324..fbe0507 100644
--- a/pyiron_vasp/vasp/output.py
+++ b/pyiron_vasp/vasp/output.py
@@ -54,7 +54,9 @@ def structure(self, atoms):
         """
         self._structure = atoms
 
-    def collect(self, directory=os.getcwd(), sorted_indices=None):
+    def collect(
+        self, directory=os.getcwd(), sorted_indices=None, es_class=ElectronicStructure
+    ):
         """
         Collects output from the working directory
 
@@ -175,7 +177,9 @@ def collect(self, directory=os.getcwd(), sorted_indices=None):
             )
             # log_dict["scf_energies"] = self.vp_new.vasprun_dict["scf_energies"]
             # log_dict["scf_dipole_moments"] = self.vp_new.vasprun_dict["scf_dipole_moments"]
-            self.electronic_structure = self.vp_new.get_electronic_structure()
+            self.electronic_structure = self.vp_new.get_electronic_structure(
+                es_class=es_class,
+            )
             if self.electronic_structure.grand_dos_matrix is not None:
                 self.electronic_structure.grand_dos_matrix[
                     :, :, :, sorted_indices, :
@@ -409,6 +413,7 @@ def parse_vasp_output(
     structure: Atoms = None,
     sorted_indices: list = None,
     read_atoms_funct: callable = read_atoms,
+    es_class=ElectronicStructure,
 ) -> dict:
     """
     Parse the VASP output in the working_directory and return it as hierachical dictionary.
@@ -440,7 +445,9 @@ def parse_vasp_output(
     output_parser.structure = structure.copy()
     try:
         output_parser.collect(
-            directory=working_directory, sorted_indices=sorted_indices
+            directory=working_directory,
+            sorted_indices=sorted_indices,
+            es_class=es_class,
         )
     except VaspCollectError:
         raise
diff --git a/pyiron_vasp/vasp/vasprun.py b/pyiron_vasp/vasp/vasprun.py
index 0aa19f7..bf0bf9e 100644
--- a/pyiron_vasp/vasp/vasprun.py
+++ b/pyiron_vasp/vasp/vasprun.py
@@ -686,7 +686,7 @@ def get_final_structure(self):
         # except (KeyError, AttributeError, ValueError):
         #     return
 
-    def get_electronic_structure(self):
+    def get_electronic_structure(self, es_class=ElectronicStructure):
         """
         Get's the electronic structure from the VASP calculation
 
@@ -694,7 +694,7 @@ def get_electronic_structure(self):
             pyiron.atomistics.waves.electronic.ElectronicStructure: The electronic structure object
 
         """
-        es_obj = ElectronicStructure()
+        es_obj = es_class()
         es_obj.kpoint_list = self.vasprun_dict["kpoints"]["kpoint_list"]
         es_obj.kpoint_weights = self.vasprun_dict["kpoints"]["kpoint_weights"]
         es_obj.eigenvalue_matrix = self.vasprun_dict["grand_eigenvalue_matrix"]
